axelor-budget/src/main/java/com/axelor/apps/budget/service/compute/BudgetLineComputeServiceImpl.java

```diff
@@ -0,0 +1,194 @@
+    Invoice invoice = Optional.ofNullable(invoiceLine).map(InvoiceLine::getInvoice).orElse(null);
+    if (invoice != null
+        && (invoice.getPurchaseOrder() != null
+            || invoice.getSaleOrder() != null
+            || invoiceLine.getPurchaseOrderLine() != null
+            || invoiceLine.getSaleOrderLine() != null)) {
+      computeMethod = this::updateBudgetLineAmountsWithPo;
+    }
+
+    updateBudgetLineAmounts(budget, amount, fromDate, toDate, defaultDate, computeMethod);
+  }
+
+  @Override
+  public void updateBudgetLineAmountsOnOrder(
+      BigDecimal amountInvoiced,
+      Budget budget,
+      BigDecimal amount,
+      LocalDate fromDate,
+      LocalDate toDate,
+      LocalDate defaultDate) {
+    if (amountInvoiced.signum() != 0) {
+      return;
+    }
+
+    ComputeMethod computeMethod = this::updateBudgetLineAmountsOnOrder;
+
+    updateBudgetLineAmounts(budget, amount, fromDate, toDate, defaultDate, computeMethod);
+  }
+
+  protected void updateBudgetLineAmounts(
+      Budget budget,
+      BigDecimal amount,
+      LocalDate fromDate,
+      LocalDate toDate,
+      LocalDate defaultDate,
+      ComputeMethod computeMethod) {
+    if (budget == null || amount.signum() == 0 || computeMethod == null) {
+      return;
+    }
+
+    if (fromDate != null && toDate != null) {
+      long totalDuration = LocalDateHelper.daysBetween(fromDate, toDate, false);
+      LocalDate date = fromDate;
+      BigDecimal prorataAmount;
+      BigDecimal missingAmount = amount;
+
+      while (!date.isAfter(toDate)) {
+        BudgetLine budgetLine = budgetLineRepository.findCurrentByDate(budget, date);
+        if (budgetLine != null && totalDuration > 0) {
+          if (!toDate.isAfter(budgetLine.getToDate())) {
+            prorataAmount = missingAmount;
+          } else {
+            long duration = LocalDateHelper.daysBetween(date, budgetLine.getToDate(), false);
+            prorataAmount =
+                amount.multiply(
+                    BigDecimal.valueOf(duration)
+                        .divide(
+                            BigDecimal.valueOf(totalDuration),
+                            AppBaseService.COMPUTATION_SCALING,
+                            RoundingMode.HALF_UP));
+            prorataAmount =
+                prorataAmount.setScale(
+                    AppBaseService.DEFAULT_NB_DECIMAL_DIGITS, RoundingMode.HALF_UP);
+            missingAmount = missingAmount.subtract(prorataAmount);
+
+            if (missingAmount.compareTo(gap) == 0) {
+              prorataAmount = prorataAmount.add(gap);
+            }
+          }
+          computeMethod.computeBudgetLineAmounts(budgetLine, prorataAmount);
+          date = budgetLine.getToDate().plusDays(1);
+        } else {
+          break;
+        }
+      }
+    } else {
+      BudgetLine budgetLine = budgetLineRepository.findCurrentByDate(budget, defaultDate);
+      computeMethod.computeBudgetLineAmounts(budgetLine, amount);
+    }
```

Comment:
this part is unreadable with way too many nested conditions, please improve this.